liesel/goose/optim.py

```diff
@@ -0,0 +1,233 @@
+    batching functionality implemented here assumes a "flat" model structure. This means
+    that this function assumes that, for all :class:`.Var` objects in your model, it
+    is valid to index their values like this::
+
+        var_object.value[batch_indices, ...]
+
+    The batching functionality also assumes that all objects that should be batched
+    are included as :class:`.Var` objects with ``Var.observed`` set to ``True``.
+
+    Examples
+    --------
+
+    We show a minimal example. First, import ``tfd``.
+
+    >>> import tensorflow_probability.substrates.jax.distributions as tfd
+
+    Next, generate some data.
+
+    >>> key = jax.random.PRNGKey(42)
+    >>> key, subkey = jax.random.split(key)
+    >>> x = jax.random.normal(key, (100,))
+    >>> y = 0.5 + 1.2 * x + jax.random.normal(subkey, (100,))
+
+    Next, set up a linear model. For simplicity, we assume the scale to be fixed to the
+    true value of 1.
+
+    >>> coef = lsl.param(jnp.zeros(2), name="coef")
+    >>> xvar = lsl.obs(jnp.c_[jnp.ones_like(x), x], name="x")
+    >>> mu = lsl.Var(lsl.Calc(jnp.dot, xvar, coef), name="mu")
+    >>> ydist = lsl.Dist(tfd.Normal, loc=mu, scale=1.0)
+    >>> yvar = lsl.obs(y, ydist, name="y")
+    >>> model = lsl.GraphBuilder().add(yvar).build_model()
+
+    Now, we are ready to run the optimization.
+
+    >>> result = gs.optim_flat(model, params=["coef"])
+    >>> result.position
+    {'coef': Array([0.5227438, 1.2980561], dtype=float32)}
+
+    We can now, for example, use ``result.model_state`` in
+    :meth:`.EngineBuilder.set_initial_values` to implement a "warm start" of MCMC
+    sampling.
+    """
+    if optimizer is None:
+        optimizer = optax.adam(learning_rate=1e-1)
+
+    n = _find_sample_size(model)
+    observed = _find_observed(model)
+    interface = gs.LieselInterface(model)
+    position = interface.extract_position(params, model.state)
+    interface._model.auto_update = False
+
+    def _batched_neg_log_prob(position: Position, batch_indices: Array | None = None):
+        if batch_indices is not None:
+            batched_observed = batched_nodes(observed, batch_indices)
+            position = position | batched_observed  # type: ignore
+
+        updated_state = interface.update_state(position, model.state)
+        return -updated_state["_model_log_prob"].value
+
+    neg_log_prob_grad = jax.grad(_batched_neg_log_prob, argnums=0)
+
+    opt_state = optimizer.init(position)
+
+    nlp_start = _batched_neg_log_prob(position, batch_indices=None)
+    abs_change: Array | float = 100.0
+    rel_change: Array | float = 100.0
+    i = 0
+
+    if batch_size is None:
+        batch_size = n
+
+    n_batches_inside = n // batch_size
+    batch_indices = jnp.arange(0, batch_size)
+    last_batch_size = n % batch_size
+    last_batch_start = n_batches_inside * batch_size
+    last_batch_end = last_batch_start + last_batch_size
+    last_batch_indices = jnp.arange(last_batch_start, last_batch_end)
```

Comment:
I think it would be better to change the batch indexes at every iteration. This usually helps to avoid local minima and improves generalization. Also, you can have the last batch of the same size as the others by resampling some data points without replacement.

```python
def _generate_batches_idx(slef, key: JaxArray, N: int, batch_size: int) -> JaxArray:
        key_permutation, key_choice = random.split(key, 2)
        batches_idx = jnp.hstack([
            jax.random.permutation(key_permutation, N),
            jax.random.choice(key_choice, N, (batch_size - N % batch_size, ))
        ])
        return jnp.array_split(batches_idx, len(batches_idx) // batch_size)
```

Reply:
Oh, so this would mean you don't split up the data into batches of a certain size, but you always draw a random sample of the batch size with replacement? So the same data points can get used twice?

Reply:
You split the data in batches of a certain size but you change the elements in each batch every iteration. It might be that N % batch_size != 0. In that case, you can resample some data points to fill the batch. It is like doing a "data augmentation".

Example:

You have 5 data points (x_1, x_2, x_3, x_4, X_5) and a batch size of 2:

Iteration 1: (x_1, x_3), (x_4, x_2), (x_5, x_3) sampled randomly
Iteration 2: (x_4, x_5), (x_2, x_1), (x_3, x_4) sampled randomly
Iteration3: ...

Reply:
That sounds good! Thanks a lot! 😊

Reply:
@wiep suggests dropping the last, non-full batch instead of filling it via resampling. This is probably more common.

Reply:
I haven't found a paper where they prove that adding some resample observations to the last batch leads to an unbiased estimator, but here are two example where they do that:

- [neon](https://neon.nervanasys.com/docs/latest/faq.html)
- [CNN Sentence](https://github.com/yoonkim/CNN_sentence/blob/master/conv_net_sentence.py) from this popular paper [Convolutional Neural Networks for Sentence Classification](https://arxiv.org/pdf/1408.5882.pdf%EF%BC%89%E3%80%82%E4%B8%8B%E9%9D%A2%E7%9C%8B%E4%B8%8B%E8%AE%BA%E6%96%87%E4%B8%AD%E7%9A%84%E6%88%AA%E5%9B%BE%EF%BC%9A)

Reply:
Thanks for the research! :)